Expose option to explicitly specify JPostal library path

vcschapp opened the pull request:
# Description

The purpose of this change is to enable using JPostal within [AWS Glue](https://docs.aws.amazon.com/glue/latest/dg/what-is-glue.html).

This unblocks "Big Data" use cases of parsing/expanding addresses within a distributed Spark environment on the AWS-managed Glue platform.

Within AWS Glue's managed Spark environment, it is pretty much impossible to override `LD_LIBRARY_PATH` in a way that makes JPostal's `System.load("jpostal")` call able to find the JPostal shared object.

## Change details

This change refactors the code base to add an option to explicitly specify the library location so the JPostal shared object can be provided to Glue driver and executor nodes via Glue's existing extra files/additional files/referenced files feature, and Glue users can load the library from the place the Glue framework stores it, which is currently `/tmp/`, at least on the latest version, Glue 5.

### Example use of new feature

```java
AddressExpander expander = AddressExpander.getInstanceConfig(Config.builder()
    .dataDir("/path/to/data")
    .libraryName("/exact/path/to/jpostal.so")
    .build()
);
...
```

The same construct is also available for `AddressParser`.

## Backward-compatibility

This change is fully backwardly-compatible. JPostal users can continue doing this:

```java
AddressExpander expander = AddressExpander.getInstance();
...
```

And this:

```java
AddressExpander expander = AddressExpander.getInstanceDataDir("/path/to/data");
...
```

As well as using the equivalent constructs for `AddressParser`.

## Incidental Changes

As part of this change, the thread-safety of global `libpostal` state is improved in the following ways.

- Global `libpostal` library initialization is done only once, by a `LibPostal` singleton class that is always initialized before  `AddressExpander` or `AddressParser`.
- All attempts to run a duplicate initialization of the main library with different configuration, *e.g.* different native library path, different data directory path, *etc.* should now be impossible because they have to run through the `LibPostal` singleton.

# Testing Done

1. ✅ DONE. Added new unit tests.
2. ✅ DONE. Tested using JPostal in a Scala-based AWS Glue 5 job.

# Example

Example usage of these new features within a working AWS Glue 5 Scala script:

```scala
import com.amazonaws.services.glue.GlueContext
import com.amazonaws.services.glue.log.GlueLogger
import com.mapzen.jpostal.{AddressExpander, AddressParser, Config};
import java.io._
import java.nio.file.{Files, Paths}
import java.util.zip.{ZipEntry, ZipInputStream}
import org.apache.spark.SparkContext
import org.apache.spark.sql.{Row, SparkSession}
import org.apache.spark.sql.api.java.UDF1
import org.apache.spark.sql.functions.udf;
import org.apache.spark.sql.types.{ArrayType, MapType, StringType, StructField}

object LibPostal {
    private val logger = new GlueLogger;
    private lazy val (expander, parser) = {
        loadSharedObjects("/tmp", Seq("libpostal.so.1.0.1", "libjpostal.so"))
        unpackData("/tmp", "data.zip", "/tmp/data")
        initJPostal("/tmp/data", "/tmp/libjpostal.so")
    }

    private def loadSharedObjects(dirPath: String, libraryNames: Seq[String]) {
        logger.info(String.format("Loading shared objects [%s] from directory [%s]..", libraryNames.mkString(", "), dirPath));
        libraryNames.foreach { libraryName => 
            val libraryPath = s"$dirPath/$libraryName";
            logger.info(String.format("Loading shared object [%s]..", libraryPath));
            System.load(libraryPath);
            logger.info(String.format("Loaded shared object [%s].", libraryPath));
        }
        logger.info(String.format("Loaded shared objects [%s] from directory [%s].", libraryNames.mkString(", "), dirPath));
    }

    private def unpackData(srcDir: String, srcZipFileName: String, dstDir: String) {
        logger.info(s"Unpacking ZIP archive [$srcZipFileName] from directory [$srcDir] to directory [$dstDir]...")
        
        val dstDirPath = Paths.get(dstDir)

        logger.info(s"Creating directory [$dstDirPath]...")
        Files.createDirectories(dstDirPath)

        val buf = new Array[Byte](8192)
        val zipFilePath = Paths.get(s"$srcDir/$srcZipFileName")
        logger.info(s"Opening ZIP archive [$zipFilePath]...")
        val zipIn = new ZipInputStream(new FileInputStream(zipFilePath.toFile))
        var numFiles: Long = 0
        var sizeCompressed: Long = 0
        var sizeUncompressed: Long = 0

        logger.info("Unpacking...")
        try {
            var entry: ZipEntry = zipIn.getNextEntry
            while (entry != null) {
                val filePath = dstDirPath.resolve(entry.getName)
                if (entry.isDirectory) {
                    Files.createDirectories(filePath)
                } else {
                    logger.info(s"Unpacking [${entry.getSize}] uncompressed bytes to [${filePath}]...")
                    Files.createDirectories(filePath.getParent)
                    val fileOut = new FileOutputStream(filePath.toFile)
                    try {
                        var len = zipIn.read(buf)
                        while (len > 0) {
                            fileOut.write(buf, 0, len)
                            len = zipIn.read(buf)
                        }
                    } finally {
                        fileOut.close()
                    }
                    numFiles += 1
                    sizeCompressed += entry.getCompressedSize
                    sizeUncompressed += entry.getSize
                }
                entry = zipIn.getNextEntry
            }
        } finally {
            zipIn.close()
        }

        logger.info(s"Unpacked [$srcZipFileName] from directory [$srcDir] to directory [$dstDir]. Got [$numFiles] files and [$sizeUncompressed] uncompressed bytes from [$sizeCompressed] compressed bytes.")
    }
    
    private def initJPostal(dataDir: String, libraryFile: String): (AddressExpander, AddressParser) = {
        val config = Config.builder()
           .dataDir(dataDir)
           .libraryFile(libraryFile)
           .build();

        logger.info(s"Initializing address expander with configuration: [$config]...")
        val expander = AddressExpander.getInstanceConfig(config)

        logger.info("Initializing address parser...")
        val parser = AddressParser.getInstanceConfig(config)

        logger.info("Finished initializing libpostal/JPostal")
    
        (expander, parser)
    }
    
    def expandAddress(address: String): Array[String] = {
        expander.expandAddress(address)
    }

    def parseAddress(address: String): Map[String, String] = {
        val parsedComponents = parser.parseAddress(address)
        
        parsedComponents
            .toSeq
            .map(c => c.getLabel -> c.getValue)
            .toMap
    }
}

class ExpandAddress extends UDF1[String, Array[String]] {
    override def call(address: String): Array[String] = {
        LibPostal.expandAddress(address)
    }
}

object ExpandAddress {
    val schema = new ArrayType(StringType, false)
}

class ParseAddress extends UDF1[String, Map[String, String]] {
    override def call(address: String): Map[String, String] = {
        LibPostal.parseAddress(address)
    }
}

object ParseAddress {
    val schema = new MapType(StringType, StringType, valueContainsNull = false)
}

/**
 * Glue expects to find an `object` named `GlueApp` in the compiled script and will
 * fail with a `ClassNotFoundException` if it doesn't find it. You can change this
 * behavior with the `--class` argument, the default value of which, in Glue 5, is
 * `GlueApp`.
 *
 * S3 JARs listed in the "Dependent JARS path" (`--extra-jars` argument) get
 * automatically placed on the JVM classpath.
 *
 * S3-hosted objects listed in the "Referenced files path" (`--extra-files` argument)
 * get automatically downloaded to `/tmp`.
 */
object GlueApp {
  def main(sysArgs: Array[String]): Unit = {
    val sparkContext: SparkContext = new SparkContext()
    val glueContext: GlueContext = new GlueContext(sparkContext)
    val sparkSession: SparkSession = glueContext.getSparkSession
    import sparkSession.implicits._ // Weird Scala thing to import extension methods like .toDF(...)`

    val expandAddressUdf = udf(new ExpandAddress, ExpandAddress.schema)
    val parseAddressUdf = udf(new ParseAddress, ParseAddress.schema)
    
    val df = Seq(
        "Quatre vingt douze Ave des Champs-Élysées",
        "1600 Pennsylvania Avenue NW, Washington, D.C. 20500, U.S."
    ).toDF("input")

    val resultDf = df
        .withColumn("expandedAddress", expandAddressUdf($"input"))
        .withColumn("parsedAddress", parseAddressUdf($"input"))

    resultDf.show(Int.MaxValue, truncate = false);
  }
}
```

vcschapp committed:
Expose option to explicitly specify JPostal library path

The purpose of this change is to enable using JPostal within AWS Glue.

This unblocks "Big Data" use cases of parsing/expanding addresses within
a distributed Spark environment on the AWS-managed Glue platform.

Within AWS Glue's managed Spark environment, it is pretty much
impossible to override `LD_LIBRARY_PATH` in a way that makes JPostal's
`System.load("jpostal")` call able to find the JPostal shared object.

This commit refactors the code base to add an option to explicitly
speclify the library location so the JPostal shared object can be
provided to Glue driver and executor nodes via Glue's existing
extra files/additional files/referenced files feature, and Glue users
can load the library from the place the Glue framework stores it, which
is currently `/tmp/`, at least on the latest version, Glue 5.

This commit is fully backwardly-compatible. JPostal users can continue
using the existing `Address[Parser|Expander].getInstance()` and
`getInstanceDataDir()` methods without any change of behavior.

vcschapp commented on src/main/java/com/mapzen/jpostal/Config.java:
```diff
@@ -0,0 +1,75 @@
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        } else if (!(o instanceof Config other)) {
+            return false;
```
Minor callout: This `instanceof` pattern syntax sugar was added in Java 14, so continuing to use it would effectively set the minimum JDK version going forward at 14.

This seems fine to me since older JDKs including 8 and 11 are no longer LTS, IIRC.

If we want to drop this and use the older and more broadly-available syntax, that's a quick fix on my side and I have no problem making it!

vcschapp reviewed (commented):
@ets Would you mind taking a look? I noticed you are the author of the most recently-merged PR (https://github.com/openvenues/jpostal/pull/44) and were made a maintainer...

@albarrentine Would you mind taking a look?

vcschapp commented on src/main/java/com/mapzen/jpostal/ExpanderOptions.java:
```diff
@@ -201,12 +201,7 @@ public static class Builder {
-            super();
-            setDefaultOptions();
-        }
+        native synchronized void setDefaultOptions();
```
This is moved into the `AddressExpander` singleton.

vcschapp commented on src/main/java/com/mapzen/jpostal/ExpanderOptions.java:
```diff
@@ -176,10 +176,6 @@ public void setRomanNumerals(boolean romanNumerals) {
     }
 
     public static class Builder {
-        static {
-            System.loadLibrary("jpostal"); // Load native library at runtime
-        }
-
```
Class loading a builder class for an options class should be lightweight and not require a native library/shared object load!

The solution in this PR is to give the `AddressExpander` singleton single responsibility over the native library load. When the singleton is created, it both loads the shared object/native library and sets the default options.

ets commented:
@vcschapp these look like nice changes.  TestAddressExpander: testEnglishExpansions is failing for me locally with your changes. Are you seeing the same?

vcschapp commented:
> @vcschapp these look like nice changes. TestAddressExpander: testEnglishExpansions is failing for me locally with your changes. Are you seeing the same?

Argh, yes I am. My apologies, I'd assumed `./gradlew assemble` would run the tests but it doesn't, so I let an error slip in between my last IDE run and my last command line run.

I'll fix it ASAP and try to push the fix.

vcschapp commented:
> > @vcschapp these look like nice changes. TestAddressExpander: testEnglishExpansions is failing for me locally with your changes. Are you seeing the same?
> 
> Argh, yes I am. My apologies, I'd assumed `./gradlew assemble` would run the tests but it doesn't, so I let an error slip in between my last IDE run and my last command line run.
> 
> I'll fix it ASAP and try to push the fix.

OK, I see what I did wrong. Dang, a somewhat further refactor is needed as I misunderstood how the default options work. I'll come back to this next week with a fix.

vcschapp committed:
Separate libpostal core initialization, fix prev. commit test fail

In this commit, I believe I have fixed the "root cause" of the problem
that ets@ commented about on the first version of my pull request to
the JPostal repo, https://github.com/openvenues/jpostal/pull/46.

In my initial commit/PR, I had misinterpreted how the builders worked,
because it's a bit counterintuitive. I had naively thought that their
`setDefaultOptions` methods were pushing global/static configuration
into libpostal and thus only had to run once, whereas the truth is they
*pull* global/static configuration from libpostal.

I looked at different ways to solve this. The ideal would be to
refactor the builders to contain boxed objects where `null` represents
the default value, and have the JNI C code check each option field and
if its null just pull the value from the corresponding libpostal global
default field at the time that parse/expand is called. However, I
rejected this option because it would result in a backwardly-
incompatible change to the builders/options.

In the end, this commit is basically putting the `setDefaultOptions`
logic back the way it was before my previous commit. However, there
is one improvement, which is that I have factored the native library
load into a single location (driven by the package-internal singleton
class `LibPostal`). This is really logically necessary because the
expander and parser both rely on the same underlying library setup
and they should not try to set it up in conflicting ways.

My improvement also allows core use of thread-unsafe global library
constructs to be used in a thread-safe manner within Java by making
it synchronized on the singleton `LibPostal` instance rather than
on the `AddressParser` and `AddressExpander` classes, since there
are two of those classes, which means having some global init
synchronized on one, and some on the other, can lead to race
conditions.

vcschapp commented:
> @vcschapp these look like nice changes. TestAddressExpander: testEnglishExpansions is failing for me locally with your changes. Are you seeing the same?

@ets, I just updated the PR to address that failing test and fix the misunderstanding I had brought in.

It's a separate commit, which you can squash or not as you see fit!

Ideally I would have refactored things to simplify and make more idiomatic Java, but I really tried hard to avoid introducing any backwardly-incompatible API changes so it's more additive.

A core improvement in the update is to centralize the `libpostal.so` native library loading and shared setup logic in one place so it can never be called twice from Java in a race condition or with conflicting configuration.

ets commented:
Thanks @vcschapp the elimination of race conditions and the new tests are appreciated!

albarrentine commented:
Looks good, @vcschapp can you add the above instructions and note about JDK versions to the README?
